Parse currency as cents

ryanwood opened the pull request:
Closes #166. A couple of notes:

 * Fully backward-compatible. Supports all existing defaults.
* Respects the precision option
* Truncates decimals passed in when `fromCents` is true.
* Added docs, TypeScript, and Flow data

Let me know what you think.

scurker commented on readme.md:
````diff
@@ -72,6 +72,15 @@ var value = currency("123.45");
+```javascript
+currency(123, { fromCents: true });               // 1.23
+currency('123', { fromCents: true });             // 1.23
+currency(123, { fromCents: true, precision: 0 }); // 0.123
+currency(123, { fromCents: true, precision: 3 }); // 123
````
This doesn't seem right, shouldn't `currency(123, { fromCents: true, precision: 0 })` be `123`? I would also expect the inverse from `precision: 3` to be `0.123`

ryanwood replied:
Correct. Sorry about that.

scurker commented on readme.md:
````diff
@@ -72,6 +72,15 @@ var value = currency("123.45");
 currency(value);    // 123.45
 ```
 
+Currency is in dollars by default but can be parsed as cents. This will respect the precision option when parsing.
````
```suggestion
Currency accepts decimal values (i.e. `1.23`) with a default precision of 2, but can accept a minor currency unit (e.g. cents in a dollar). This will respect the precision option when parsing.
```

scurker commented on readme.md:
```diff
@@ -130,6 +139,9 @@ When implementing a currency that implements rounding, setting the increment val
 Formats number groupings using the Indian Numbering System, i.e. `10,00,000.00`
 
+`fromCents` *default*: `false`<br/>
+The amount argument is parsed as cents instead of dollars.
```
```suggestion
Parse the amount value as a minor currency unit (e.g. cents in a dollar) instead of dollars.
```

scurker commented on test/test.js:
```diff
@@ -449,4 +449,20 @@ test('should handle max safe integer', t => {
\ No newline at end of file
+});
+
+test('should allow creation from cents', t => {
```
Some of the test assertions can be simplified to use `intValue` and `value` properties instead of strings. Those values should be the same for `fromCent` values and regular values of the same precision:

```js
let decimals = currency(1.23);
let fromCents = currency(123, { fromCents });
decimals.value === fromCents.value; // true
decimals.intValue === fromCents.intValue; // true
```

I would also like to see these tests broken out so it's easier to tell when a specific assertion has broken:

```js
it('should parse cents from a string when using fromCents option', t => {
  let c1 = currency('123', { fromCents })
  t.is(c1.value, 1.23)
  t.is(c1.intValue, 123)
})

it('should parse cents from a number when using fromCents option', t => { ... })
it('should truncated decimals from a string when using fromCents option', t => { ... })
it('should truncated decimals from a number when using fromCents option', t => { ... })
```

Other tests should cover the default value.

scurker commented:
Good job on keeping the size small! ❤️ 

**Size Change:** +184 B (2%) 

**Total Size:** 7.4 kB

| Filename | Size | Change | |
|:--- |:---:|:---:|:---:|
| `dist/currency.es.js` | 2.04 kB | +55 B (2%) |  |
| `dist/currency.js` | 2.06 kB | +54 B (2%) |  |
| `dist/currency.min.js` | 1.13 kB | +20 B (1%) |  |
| `dist/currency.umd.js` | 2.18 kB | +55 B (2%) |  |

ryanwood commented:
Thanks, @scurker. All of your comments have been addressed. I've rebased on `next` and squashed into a single commit. Let me know if you want any other changes.

scurker commented:
Thanks @ryanwood, this looks great! One last change is to remove the `package-lock.json` change and this should be good to go. Don't worry about the extra commit, I can just squash on merge.

ryanwood commented:
Sorry about that. I'm not sure how that package-lock.json got updated. I also realized that I changed the base of the PR to next but didn't rebase onto next. I removed the change to package-lock.json and rebased cleanly on next with a single commit. It should be good to go now.
